libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,291 @@
+        self
+    }
+
+    /// Set the parent of a hook
+    /// Will return EINVAL if set upon an TC_EGRESS/TC_INGRESS/TC_EGRESS|TC_INGRESS hook
+    ///
+    /// Must be set on a TC_CUSTOM hook
+    /// Current acceptable values are TC_H_CLSACT for maj, and
+    /// TC_H_MIN_EGRESS or TC_H_MIN_INGRESS for min
+    pub fn parent(&mut self, maj: u32, min: u32) -> &mut Self {
+        /* values from libbpf.h BPF_TC_PARENT() */
+        let parent = (maj & TC_H_MAJ_MASK) | (min & TC_H_MIN_MASK);
+        self.hook.parent = parent;
+        self
+    }
+
+    /// Set whether this hook should replace an existing hook
+    ///
+    /// If replace is not true upon attach, and a hook already exists
+    /// an EEXIST error will be returned from attach()
+    pub fn replace(&mut self, replace: bool) -> &mut Self {
+        if replace {
+            self.opts.flags = BPF_TC_F_REPLACE;
+        } else {
+            self.opts.flags = 0;
+        }
+        self
+    }
+
+    /// Set the handle of a hook.
+    /// If unset upon attach, the kernel will assign a handle for the hook
+    pub fn handle(&mut self, handle: u32) -> &mut Self {
+        self.opts.handle = handle;
+        self
+    }
+
+    /// Set the priority of a hook
+    /// if unset upon attach, the kernel will assign a priority for the hook
+    pub fn priority(&mut self, priority: u32) -> &mut Self {
+        self.opts.priority = priority;
+        self
+    }
+
+    /// Query a hook to inspect the program identifier (prog_id)
+    pub fn query(&mut self) -> Result<u32> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_query(&self.hook as *const _, &mut opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(opts.prog_id)
+        }
+    }
+
+    /// Attach the TcHook so that the program starts processing
+    ///
+    /// Once the hook is processing, changing the values will have no effect
+    /// unless the hook is attach()'d again (replace(true) being required)
+    ///
+    /// Users can create a second hook by changing the handle, the priority
+    /// or the attach_point and calling the attach() method again.
+    /// Beware doing this.  It might be better to Copy the TcHook and change
+    /// the values on the copied hook for easier detach()
+    ///
+    /// NOTE: Once a TcHook is attached, it, and the maps it uses,
+    ///       will outlive the userspace application that spawned them
+    ///       Make sure to detach if this is not desired
+    pub fn attach(&mut self) -> Result<Self> {
+        self.opts.prog_id = 0;
+        let err =
+            unsafe { libbpf_sys::bpf_tc_attach(&self.hook as *const _, &mut self.opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(*self)
+        }
+    }
+
+    /// Detach a TcHook
```

Comment:
Can you clarify  this as "Detach a filter from a TcHook"?

Reply:
Resolved